scrape/target.go

```diff
@@ -366,13 +366,21 @@ type bucketLimitAppender struct {
 
 func (app *bucketLimitAppender) AppendHistogram(ref storage.SeriesRef, lset labels.Labels, t int64, h *histogram.Histogram, fh *histogram.FloatHistogram) (storage.SeriesRef, error) {
 	if h != nil {
-		if len(h.PositiveBuckets)+len(h.NegativeBuckets) > app.limit {
-			return 0, errBucketLimit
+		for len(h.PositiveBuckets)+len(h.NegativeBuckets) > app.limit {
+			lastSum := len(h.PositiveBuckets) + len(h.NegativeBuckets)
+			h = h.ReduceResolution(h.Schema - 1)
+			if len(h.PositiveBuckets)+len(h.NegativeBuckets) == lastSum {
```

Comment:
Isn't it possible that we have two buckets so far apart that one reduction doesn't merge them, but two reductions  does? In my PR in Mimir I actually do reductions until I reach the minimal schema (-4) instead of checking whether the number of buckets has decreased like here.

Reply:
In this line I want to express that, we will reduce the resolution until we can not merge more buckets by reducing resolutions any more.

We can take the histogram in unit test `TestBucketLimitAppender` as example.

```
	example := histogram.Histogram{
		Schema:        0,
		Count:         21,
		Sum:           33,
		ZeroThreshold: 0.001,
		ZeroCount:     3,
		PositiveSpans: []histogram.Span{
			{Offset: 0, Length: 3},
		},
		PositiveBuckets: []int64{3, 0, 0},
		NegativeSpans: []histogram.Span{
			{Offset: 0, Length: 3},
		},
		NegativeBuckets: []int64{3, 0, 0},
	}
``` 

When schema is `0`, the positive and negative spans will be `{Offset: 0, Length: 3}` and `{Offset: 0, Length: 3}`, bucket counts is **6**.

When scheme is reduced to `-1`, the positive and negative spans will be `{Offset: 0, Length: 2}` and `{Offset: 0, Length: 2}`, bucket counts is **4**.

When schema is reduced to `-2`, the positive and negative spans will be `{Offset: 0, Length: 2}` and `{Offset: 0, Length: 2}`, bucket counts is still **4**.

So in this case, we can know that the bucket counts will not decrease even we reduce the resolution to -3, -4. So we can just break the loop.

@krajorama What do you think is the best way to break the loop of reducing the resolution?

Reply:
Hi, 

for example:
```
	bigGap := histogram.Histogram{
		Schema:        0,
		Count:         21,
		Sum:           33,
		ZeroThreshold: 0.001,
		ZeroCount:     3,
		PositiveSpans: []histogram.Span{
			{Offset: 1, Length: 1}, // in (1, 2]
			{Offset: 2, Length: 1}, // in (8, 16]
		},
		PositiveBuckets: []int64{1, 0}, // 1, 1
	}
```

Reducing the above histogram by one resolution does not reduce the number of buckets as the observations will be in
```
(1,4] and (4, 16]
```

However in the next step you get
```
(1, 16]
```

So I'd expect:
```
		{
			h: bigGap,
			limit: 1,
			expectError: false,
			expectBucketCount: 1,
			expectSchema: -2,
		},
```

Basically you have to continue until there are too many buckets and you haven't reached minimum schema.

e.g.:
```
	if h != nil {
		for len(h.PositiveBuckets)+len(h.NegativeBuckets) > app.limit {
			if h.Schema == -4 {
				return 0, errBucketLimit
			}
			h = h.ReduceResolution(h.Schema - 1)
		}
	}
	if fh != nil {
		for len(fh.PositiveBuckets)+len(fh.NegativeBuckets) > app.limit {
			if fh.Schema == -4 {
				return 0, errBucketLimit
			}
			fh = fh.ReduceResolution(fh.Schema - 1)
		}
	}
```

Reply:
@krajorama I already change as you suggested.